app/services/order_workflow.rb

```diff
@@ -35,9 +43,31 @@ def advance_order!(options)
     end
   end
 
+  def advance_order_one_step
+    tries ||= 3
+    order.next
+  rescue ActiveRecord::StaleObjectError
+    retry unless (tries -= 1).zero?
+    false
+  end
+
   def after_transition_hook(options)
     if order.state == "delivery"
       order.select_shipping_method(options[:shipping_method_id]) if options[:shipping_method_id]
     end
+
+    persist_all_payments if order.state == "payment"
   end
+
+  # When a payment fails, the order state machine rollbacks all transactions
```

Comment:
just thinking out loud. If that's not what we want we'll need to split up that single transaction at some point so that a rollback while persisting payment doesn't affect the order changes. Definitely not today. Removing that callback is a lot already.

Reply:
yeah, I'd love to do that, but I think it's state_machine that does the rollback by default if the workflow step returns false.
btw, that's why I brought Checkout.rb from spree, because I was trying to do exactly that. But I dont know enough about the state_machine gem... the other problem is that we may want the rollback to happen for other objects other than the payments...

Reply:
yes, it feels like heart surgery. Overall, I'm a bit concerned about our current status where things are partially implemented in the Spree order state machine, plus patched on our checkout controller and our services. I guess I just need to believe in us while we simplify our customizations up until the day we can question the state machine's implementation.

Reply:
yeah, I see what you mean, but we are improving!